Implement `useInfiniteScroll` hook

orballo opened the pull request:
<!--
Thanks for your pull request 😊. Note that not following the template might
result in your issue being closed

Please make sure you're familiar with and follow the instructions in the
contributing guidelines found in the
https://docs.frontity.org/contributing/code-contributions.

If you're new to contributing to open source projects, you might find this free
video course helpful: http://kcd.im/pull-request

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

**What**:

<!-- What changes are being made? (What feature/bug is being fixed here?) -->
Implements the `useArchiveInfiniteScroll`, `usePostsInfiniteScroll` and `useInfiniteScroll` hooks.

**Why**:

<!-- Why are these changes necessary? -->
To facilitate the implementation of infinite scroll to Frontity users.

**How**:

This are the types of the current implementations.
```ts
type UseInfiniteScroll = (options: {
  currentLink: string;
  nextLink?: string;
  fetchInViewOptions?: IntersectionOptions;
  routeInViewOptions?: IntersectionOptions;
}) =>
  | {
      supported: false;
      routeRef?: undefined;
      fetchRef?: undefined;
      routeInView?: undefined;
      fetchInView?: undefined;
    }
  | {
      supported: true;
      routeRef: (node?: Element) => void;
      fetchRef: (node?: Element) => void;
      routeInView: boolean;
      fetchInView: boolean;
    };
```
```ts
type UseArchiveInfiniteScroll = (options?: {
  limit?: number;
  active?: boolean;
}) => {
  pages: {
    key: string;
    link: string;
    isLast: boolean;
    Wrapper: React.FC<Connect<Source & Router>>;
  }[];
  isLimit: boolean;
  isFetching: boolean;
  fetchNext: () => Promise<void>;
};
```
```ts
type UsePostTypeInfiniteScroll = (options?: {
  limit?: number;
  active?: boolean;
  archive?: string;
  fallback?: string;
}) => {
  posts: {
    key: string;
    link: string;
    isLast: boolean;
    Wrapper: React.FC<Connect<Source & Router>>;
  }[];
  isLimit: boolean;
  isFetching: boolean;
  fetchNext: () => Promise<void>;
};
```
```ts
type Wrapper = (link: string) => React.FC<Connect<Source & Router>>;
```

Examples of how they could be used:
```tsx
import useArchiveInfiniteScroll from "@frontity/hooks/use-archive-infinite-scroll";

const Archive = () => {
  const { pages, isLimit, isFetching, fetchNext } = useArchiveInfiniteScroll();

  return (
    <div>
      {/* <ArchivePage /> would be a custom component created by the developer. */}
      {pages.map(({ key, link, isLast, Wrapper }) => (
        <Wrapper key={key}>
          <ArchivePage link={link} />
          {!isLast && <hr />}
        </Wrapper>
      ))}
      {/* <Loading /> would also be a custom component created by the developer. */}
      {isFetching && <Loading />}
      {isLimit && <button onClick={fetchNext}>Load Next Page</button>}
    </div>
  );
};
```
```tsx
import usePostTypeInfiniteScroll from "@frontity/hooks/use-post-type-infinite-scroll";

const PostTypeList = () => {
  const { state } = useConnect();
  
  const data = state.source.get(state.router.link);

  const { posts, isLimit, isFetching, fetchNext } = usePostTypeInfiniteScroll({
    active: !!data.isPost
  });

  return (
    <div>
      {/* <PostType /> would be a custom component created by the developer. */}
      {posts.map(({ key, link, isLast, Wrapper }) => (
        <Wrapper key={key}>
          <PostType link={link} />
          {!isLast && <hr />}
        </Wrapper>
      ))}
      {/* <Loading /> would also be a custom component created by the developer. */}
      {isFetching && <Loading />}
      {isLimit && <button onClick={fetchNext}>Load Next Page</button>}
    </div>
  );
};
```
<!-- How were these changes implemented? -->

**Tasks**:

<!-- Have you done all of these things?  -->

<!-- To check an item, place an "x" in the box like so: "- [x] Documentation"
Strikethrough each line that's irrelevant to your changes using ~, for example: "- [x] ~Code~"
If needed, add the reason: "- [x] ~Documentation~: Internal code, not documented." -->

- [x] Code
- [x] TypeScript
- [x] Unit tests
- [x] TypeScript tests
- [x] Changesets (one for the new `updateState` action in router,  one for the infinite scroll itself and one for the fix in `@frontity/types`)

**To be done in a later PR**:
- End to end tests
- Documentation
<!-- Changesets are necessary if your changes should release any packages.Run
`npx changeset` to create a changeset. If there is a Feature Discussion for
this, please add a link: https://community.frontity.org/c/33 -->

**Irrelevant**:
- ~Update starter themes~
- ~Update other packages~
- ~Community discussions~

<!-- Feel free to add any additional comments. -->

Commit:
wip

Commit:
initial implementation of use-infinite-scroll

Commit:
wip

Commit:
add limit to useInfiniteScroll

Commit:
wip, implements useArchiveInfiniteScroll

Commit:
revert changes in mars theme

Commit:
revert changes in mars theme

Commit:
[wip] new iteration on archive infinite scroll

Commit:
[wip] yet another iteration on archive infinite scroll

michalczaplinski committed:
Merge branch 'dev' into use-infinite-scroll

Commit:
implements useArchiveInfiniteScroll

Commit:
Merge branch 'use-infinite-scroll' of github.com:frontity/frontity into use-infinite-scroll

Commit:
wip

Commit:
Merge branch 'use-in-view' into use-infinite-scroll

Commit:
wip

Commit:
Merge branch 'dev' into use-infinite-scroll

Commit:
implement clone and refactor infinite scroll

Commit:
wip

Commit:
wip

Commit:
wip

Commit:
fixes limit in archive

Commit:
wip

Commit:
Merge branch 'use-in-view' into use-infinite-scroll

Commit:
improves limit in archive and fixes renders on fetchNext

Commit:
limits working in posts and archive

Commit:
fixes types

orballo commented:
@luisherranz Could you please have a look at this? Let me know of any changes or doubts on the API.

It still needs some adjustments and cleaning but everything seems to be working as expected. I'll be working on that and implementing tests and e2e.

luisherranz commented:
Sure, I'll take a look 🙂

Commit:
Merge branch 'dev' into use-infinite-scroll

luisherranz commented:
Sorry I haven't been able to check this today. I'll try to do it this weekend.

orballo commented:
@luisherranz I was thinking that maybe `usePostsInfiniteScroll` should take `context` option and `forceContext` option. So, what I have in mind are these three scenarios:

1. I want to use infinite scroll only if the context is set  in `state.router.state.context`.
2. I want to use any context set in `state.router.state.context` and use `options.context` as default (i.e. SSR).
3. I want to use my `options.context` no matter if the `state.router.state.context` is already set, so I can use a custom route as context like `@related-articles/:id`.

Other option would be not to use `options.context` as default but always forcing it, so that `options.context` is used only to override any previous context set, and the default context would be always `/`. If somebody wants the first scenario can force it programaticaly in the component where they are calling `usePostsInfiniteScroll`.

Let me know what do you think.

Commit:
fixes pages in usePostsInfiniteScroll

luisherranz commented:
I will have to go over `usePostsInfiniteScroll` to answer you on that. I only had time to check `useInfiniteScroll` and `useArchiveInfiniteScroll`. I'll try to do it this morning.

---

I really like it so far. And only have one suggestion: Why don't you create the `Wrapper` on demand for each page item with `link` embedded?

```ts
const Wrapper = (link: string): Wrapper =>
  connect(({ state, children }) => {
    // ...
  });
```

And then

```ts
// Map every link to its DIY object.
const pages = links.map((link) => ({
  key: link,
  link: link,
  isLastPage:
    link === last.link || (link === links[links.length - 2] && !last.isReady),
  Wrapper(link),
}));
```

Apart from that, I don't like the idea of using `actions.router.set` to replace `state.router.state`. Once we have hooks, people may subscribe to that action to do something when the URL has changed, and if we start doing this, people would have to check if the URL was the same or not.

So could you please do a PR to create `actions.router.updateState` and switch to that? Internally it does a `window.history.replaceState` with the current link and new `state`.

As soon as we have hooks, I'll create one to silently abort an `actions.router.set` if it has the same `link` to avoid causing problems to people subscribing to its executions.

luisherranz commented on packages/connect/src/index.js:
```diff
@@ -1,5 +1,5 @@
 export { observe, unobserve } from "./observer";
-export { observable, isObservable, raw } from "./observable";
+export { observable, isObservable, raw, clone } from "./observable";
```
Could you please move `clone` to `tiny-router`? Sorry I think I probably told you to put it here but I don't want this to be a public API of `connect` yet. I want to discuss it with Miklos and see if how this can be avoided.

luisherranz commented on packages/hooks/use-archive-infinite-scroll.tsx:
```diff
@@ -0,0 +1,139 @@
+import WpSource from "@frontity/wp-source/types";
+import TinyRouter from "@frontity/tiny-router/types";
+
+type Wrapper = React.FC<Connect<WpSource & TinyRouter, { link: string }>>;
```
Are you using the types of `WpSource` and `TinyRouter` instead of `Source` and `Router` for something in particular?

orballo replied:
Just because I tend to use the types of the packages I'm working with. But in this case I guess we could use the generics packages, so different implementations are compatible?

luisherranz replied:
Yes, it should work with any theme or source so it's better if we use the general ones.

luisherranz commented on packages/hooks/use-posts-infinite-scroll.tsx:
```diff
@@ -0,0 +1,242 @@
+  );
+});
+
+const usePostsInfiniteScroll: UsePostsInfiniteScroll = (options) => {
```
Could you please rename this to `usePostTypeInfiniteScroll` (singular)? I want to start using the same terms everywhere, and we are going to start using `postType` for custom post types, including posts and pages. Thanks!

luisherranz commented on packages/tiny-router/src/actions.ts:
```diff
@@ -1,5 +1,5 @@
 import TinyRouter from "../types";
-import { warn } from "frontity";
+import { warn, clone } from "frontity";
```
Just import `clone` from Ramda here.

orballo commented:
> I really like it so far. And only have one suggestion: Why don't you create the `Wrapper` on demand for each page item with `link` embedded?
> 
> ```ts
> const Wrapper = (link: string): Wrapper =>
>   connect(({ state, children }) => {
>     // ...
>   });
> ```
> 
> And then
> 
> ```ts
> // Map every link to its DIY object.
> const pages = links.map((link) => ({
>   key: link,
>   link: link,
>   isLastPage:
>     link === last.link || (link === links[links.length - 2] && !last.isReady),
>   Wrapper(link),
> }));
> ```

This forces rerenders of the whole list because I need to recreate every `Wrapper` when a new link is added to the list.

I might have miss something here. So if you have an idea on how to do that implementation, that's something worth trying again.

luisherranz commented:
You can use a `memoize` function to cache the result.

luisherranz commented:
After a call we have decided to go with this logic to decide which `archive` is going to be used for the postType infinite scroll:

![image](https://user-images.githubusercontent.com/3305402/82791095-716fd600-9e6d-11ea-8541-b7fcc20a2097.png)

Drawing: https://excalidraw.com/#json=6498078429282304,3HeDJQH8qiEJG4xFZjUfNg

orballo commented:
There are a couple of things that need to be changed regarding `usePostTypeInfiniteScroll`:

1. Rename the option/state `context` to `archive`.
2. Move the infinite scroll state from `state.router.state` to a namespace like `state.router.state.infiniteScroll`.
3. Create a `state.router.previous` so that the archive can be guessed from there.

Commit:
adds tests for useInfiniteScroll

orballo commented:
@luisherranz I was going to implement the changes on the namespace, and I realized that to prevent overriding the state I need to do, in my case, something like this:

```ts
actions.router.set(currentLink, {
  method: "replace",
  state: {
    ...state.router.state,
    infiniteScroll: {
      ...state.router.state.infiniteScroll,
      links,
    },
  },
});
```

This looks to me very convoluted and prone to errors and overriding necessary state by other packages or developers that might not be thinking about preventing that. Maybe we should go with a different API for `actions.router.set` like:

```ts
actions.router.set("/", {
  method: "replace",
  state: {
    namespace: "infiniteScroll",
    data: {
      ...state.router.state.infiniteScroll,
      links,
   },
  },
});
``` 
or
```ts
actions.router.set("/", {
  method: "replace",
  namespace: "infiniteScroll",
  state: {
      ...state.router.state.infiniteScroll,
      links,
   },
});
```

What do you think? Maybe you can suggest another approach.

luisherranz commented:
That is going to be managed with `actions.router.updateState` so what about doing a merge?

Similar to the React' `setState` in classes.

```js
actions.router.updateState({
  infiniteScroll: {
    links: [ updated ]
  }
});
```

I'm not super fan to be honest, because this may not be what the developer wants or expects. The first example, even when it looks convoluted, gives more control to the caller to delete everything or part of the current state.

orballo commented:
Yep, I thought about a merge too and it looks like a good idea, but it can be misleading. If you think the first case is not that awful, we can just let it be that way, at least for the moment?

One question, just curious, I'm thinking also about when we need to change the state with a `set` and some other package needs to do it's changes too. That other package would be listening to a hook in `actions.router.set` and then fire it's own `actioins.router.updateState` I guess, right? Or, I just remembered, you mentioned something like a `filter` that maps the params before executing the action?

Commit:
moves browser state to own namespace

Commit:
uses @frontity/source and @frontity/router types instead of @frontity/wp-source and @frontity/tiny-router

Commit:
removes clone from @frontity/connect

Commit:
renames usePostsInfiniteScroll to usePostTypeInfiniteScroll

Commit:
memoizes wrappers to prevent rerenderings and avoid passing  manually

orballo commented:
I'm going to list all the changes so they are just in one post:
- [x] Move state to `infiniteScroll` namespace.
- [x] Rename `usePostsInfiniteScroll` to `usePostTypeInfiniteScroll`.
- [x] Create `Wrapper` on demand using a memoize function to prevent rerenders.
- [x] Remove `clone` from `@frontity/connect`.
- [x] Use `@frontity/source` and `@frontity/router` types instead of `@frontity/wp-source` and `@frontity/tiny-router`.
- [x] Rename `context` to `archive` in `usePostTypeInfiniteScroll`.
- [x] Create a `state.router.previous` field holding the previous `state.router.link`.
- [x] Create the action `actions.router.updateState` to update the browser state without changing the route.
- [x] Use the action `actions.router.updateState` in the infinite scroll hooks.
- [x] Implement the new logic for `usePostTypeInfiniteScroll`.

Commit:
rename context to archive

Commit:
implements state.router.previous

Commit:
implements action.router.updateState

Commit:
migrate to actions.router.updateState

Commit:
fixes direct manipulation of state.router.state

orballo commented:
@luisherranz I'm working on the new logic for `usePostTypeInfiniteScroll` and I don't think it's possible to check if the current post type has a related archive. Is there a way to do that? Am I missing something?

luisherranz commented:
> That other package would be listening to a hook in actions.router.set and then fire it's own actions.router.updateState I guess, right? Or, I just remembered, you mentioned something like a filter that maps the params before executing the action?

Yes, they could do that, but in that case it would make sense to just modify the args. `hooks` are going to be something like this:

```js
export default {
  state: {
    hooks: {
      actions: {
        myPackage: {
          addStuffToRouterState: {
            path: "actions.router.set",
            library: "myPackage.addStuffToRouterState",
            priority: 5,
          },
        },
      },
    },
  },
  libraries: {
    hooks: {
      actions: {
        myPackage: {
          addStuffToRouterState: ({ args }) => {
            const [link, options] = args;
            options.state = {
              ...options.state,
              myStuff: {
                // ...
              },
            };
          },
        },
      },
    },
  },
};
```

> I don't think it's possible to check if the current post type has a related archive. Is there a way to do that? Am I missing something?

You're right. Right now you could check the settings of `wp-source`, but in the future, with other source packages, that information is not going to be there because postTypes will work without settings. I guess we would need a field called `archive` in each postType.

We can solve it manually, but using `options.archive` on the hook of each postType, but that would break the nice `state.router.previous` check with the current behavior. So I guess we finally need another option to make `options.archive` overwrite `state.router.previous`.

What about having:

```ts
interface Options {
  /**
   * If infinite scroll should be activated or not. Useful for sharing
   * the same component between posts and pages.
   */
  active: boolean;

  /**
   * The link of an archive to be used. It used in all circumstances,
   * even with `actions.router.set` contains an archive link.
   */
  archive: string;

  /**
   * The link of an archive to be used when other archive links are not
   * present, like the link specified on `actions.router.set` or
   * the link in `state.router.previous`.
   */
  fallback: string;
}
```

What do you think?

![image](https://user-images.githubusercontent.com/3305402/82880684-0abaed00-9f3f-11ea-8d1f-e2011f4ac537.png)


Drawing: https://excalidraw.com/#json=5687473816469504,KhkqMA8iJli8lYzK-pceDA

orballo commented:
It looks good to me. I'll implement it that way. I'll let you know if I run into something else.

Commit:
update useInfiniteScroll tests

Commit:
[wip] adds tests for useArchiveInfiniteScroll

Commit:
adds tests for useArchiveInfiniteScroll

Commit:
remove console logs

Commit:
rename isLastPage to isLast

Commit:
implement new logic in usePostTypeInfiniteScroll

Commit:
removes unnecessary dependencies

Commit:
fixes dependencies in mars-theme

Commit:
[wip] adds tests for usePostTypeInfiniteScroll

Commit:
[wip] adds tests for usePostTypeInfiniteScroll

Commit:
[wip] adds tests for usePostTypeInfiniteScroll

Commit:
[wip] adds tests for usePostTypeInfiniteScroll

Commit:
adds tests for usePostTypeInfiniteScroll

orballo commented:
@luisherranz do you think we should implement `options.active` in `useArchiveInfiniteScroll` too? I can't think of any scenario, but it might be useful for somebody to be able to stop the infinite scroll behavior at any point 🤔

luisherranz commented:
Yes, let's make both consistent 🙂

Commit:
implements options.active for useArchiveInfiniteScroll

orballo commented:
@luisherranz one more thing, I'm not sure how to call the fields `posts` (for `usePostTypeInfiniteScroll`) and `pages` (for `useArchiveInfiniteScroll`) in the return values. I was thinking maybe `list`? `entities`?

Commit:
adds types tests

luisherranz commented:
To be honest, I think that semantically `pages` in `useArchiveInfiniteScroll` and `posts` in `usePostTypeInfinteScroll` make the more sense to me.

I don't like `entities` because they don't relate to pages, only to posts or taxonomies.

I guess we could use `list` and let the user decide what to to use:

```js
// useArchiveInfiniteScroll
list.map((page) => {
  // ...
});

// usePostTypeInfinteScroll
list.map((post) => {
  // ...
});
```

versus

```js
// useArchiveInfiniteScroll
pages.map((page) => {
  // ...
});

// usePostTypeInfinteScroll
posts.map((post) => {
  // ...
});
```

Up to you, but I wouldn't mind going with `pages` and `posts` either because they belong to two different hook, one to do infinite scroll of pages and the other to do infinite scroll of posts.

Commit:
changesets

orballo commented:
Oh, I forgot to reply. I decided to keep `posts` and `pages` as you suggested.

luisherranz commented:
Let me know once this is ready for a beta release.

orballo commented:
It's ready.

luisherranz committed:
Don't create new history entry for the same link

luisherranz committed:
Avoid importing react-intersection-observer

luisherranz committed:
Switch to markdown links

luisherranz committed:
Add TSDocs to useInfiniteScroll

luisherranz committed:
Add TSDocs to useArchiveInfiniteScroll

luisherranz committed:
Add className to Wrapper to allow styling it

Some people may want to change the CSS of the Wrapper. If we add
className they can do so with with either `styled` or `css`.

luisherranz committed:
Add TSDocs to usePostTypeInfiniteScroll

luisherranz committed:
Merge branch 'dev' into use-infinite-scroll

luisherranz commented on packages/hooks/use-infinite-scroll.ts:
```diff
@@ -0,0 +1,193 @@
+
+      links.push(nextLink);
+
+      actions.router.updateState({
+        ...state.router.state,
+        infiniteScroll: {
+          ...state.router.state.infiniteScroll,
+          links,
+        },
+      });
```
My only question about your implementation is that if we should wait to update the state with the new link until the data for that link is ready.

I guess there is no problem, as the scroll is always going to be above the new fetched page/post when it's not ready. Did you think about that?

orballo replied:
If it is not ready, there is no ref catching the scroll to update the link. (At least in the current implementations) So maybe this is a valid concern if we create some other hook that uses `useInfiiniteScroll` and works in a different way.
However, even if the link is set before ready, the theme should be prepared for that.

luisherranz committed:
Merge branch 'dev'

github-actions[bot] committed:
Version Packages

luisherranz committed:
Add c4rl0sbr4v0 to the changelog

Because he added the changeset without using the correct git author (https://github.com/frontity/frontity/pull/454/commits/de6c7fe6e43510dddf26e8afbe1ef05ac5ab379c).

luisherranz committed:
Update comscore and google-tag-manager links

Because it's a major release and we have to do those manually.

luisherranz committed:
Merge pull request #457 from frontity/changeset-release/master

Version Packages

luisherranz committed:
Merge branch 'dev' into use-infinite-scroll

luisherranz committed:
Fix router types test

luisherranz reviewed (approved):
I haven't checked the tests, but the code looks great to me.

I have only one question/doubt, I have added it below. I guess we'll see if that is a problem once we implement this in real apps.

We're going to do a beta release 🙂

luisherranz committed:
Bypass TypeScript in some unit tests

Types were failing so I have added some "as any" to be able to release the
beta version. @orballo please take a look whenever you continue with this.

luisherranz commented:
Types were failing so I have added some "as any" to be able to release the beta version in https://github.com/frontity/frontity/pull/429/commits/058f65db1f85bc59becccedc890719ee1c0753bc. Please take a look whenever you continue with this.

luisherranz commented:
I have published the packages in beta. You can install them using these tags:

- `frontity@1.8.1-infinite-scroll-beta.0`
- `@frontity/core@1.7.1-infinite-scroll-beta.0`
- `@frontity/hooks@2.1.0-infinite-scroll-beta.0`
- `@frontity/tiny-router@1.3.0-infinite-scroll-beta.0`

The branch for this beta releases is [`infinite-scroll-beta`](https://github.com/frontity/frontity/tree/infinite-scroll-beta). If you need to add additional changes, open a PR on that branch.

luisherranz commented:
@SantosGuillamot @orballo @javiersefer: you can now test it out in your themes. There is an example of how to use it in the opening post of this PR.here is an example of how to use it in the opening post of this PR.
